File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
    _, balanced := depthBalance(root)
    return balanced
}

func depthBalance(root *TreeNode) (int, bool) {
    if root == nil {
        return 0, true
    }
    depthLeft, leftBalanced := depthBalance(root.Left)
    depthRight, rightBalanced := depthBalance(root.Right)
    if !leftBalanced || !rightBalanced || 1 < abs(depthLeft - depthRight) {
        return -1, false
    }
    return 1 + max(depthLeft, depthRight), true
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        if 1 < abs(self.depth(root.left) - self.depth(root.right)):
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)
    
    def depth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        return 1 + max(self.depth(root.left), self.depth(root.right))
